src-kmerge/KMerge/LoserTree.hs

```diff
@@ -35,69 +36,76 @@ placeholder = unsafeCoerce ()
 --
 -- The second half of a pair is the winner value (only losers are stored in the tree).
 --
-newLoserTree :: forall a m. (PrimMonad m, Ord a) => [a] -> m (MutableLoserTree (PrimState m) a, Maybe a)
-newLoserTree [] = do
-    ids <- newPrimArray 0
-    arr <- newSmallArray 0 placeholder
-    sizeRef <- newPrimVar 0
-    holeRef <- newPrimVar 0
-    return $! (MLT sizeRef holeRef ids arr, Nothing)
-newLoserTree [x] = do
-    ids <- newPrimArray 0
-    arr <- newSmallArray 0 placeholder
-    sizeRef <- newPrimVar 0
-    holeRef <- newPrimVar 0
-    return $! (MLT sizeRef holeRef ids arr, Just x)
-newLoserTree xs0 = do
+newLoserTree :: forall a m. (PrimMonad m, Ord a) => NonEmpty a -> m (MutableLoserTree (PrimState m) a, a)
+newLoserTree zs0@(z :| zs) = do
     -- allocate array, we need one less than there are elements.
     -- one of the elements will be the winner.
-    ids <- newPrimArray  (len - 1)
-    setPrimArray ids 0 (len - 1) (-1)
-    arr <- newSmallArray (len - 1) placeholder
-
-    loop ids arr (len - 1) xs0
+    ids <- newPrimArray  len
+    arr <- newSmallArray len placeholder
+    case zs of
+      [] -> do
+        sizeRef <- newPrimVar 0
+        holeRef <- newPrimVar 0
+        return $! (MLT sizeRef holeRef ids arr, z)
+      _ -> do
+        setPrimArray ids 0 len (-1)
+        loop ids arr len zs0
   where
-    !len = length xs0
+    !len = length zs
 
-    loop :: MutablePrimArray (PrimState m) Int -> SmallMutableArray (PrimState m) a -> Int -> [a] -> m (MutableLoserTree (PrimState m) a, Maybe a)
-    loop !_   !_   !_   []     = error "should not happen"
-    loop  ids  arr  idx (x:xs) = do
+    loop :: MutablePrimArray (PrimState m) Int -> SmallMutableArray (PrimState m) a -> Int -> NonEmpty a -> m (MutableLoserTree (PrimState m) a, a)
+    loop  ids  arr  idx (x :| xs) = do
         sift ids arr (parentOf idx) (parentOf idx) x idx xs
 
-    sift :: MutablePrimArray (PrimState m) Int -> SmallMutableArray (PrimState m) a -> Int -> Int -> a -> Int -> [a] -> m (MutableLoserTree (PrimState m) a, Maybe a)
+    sift :: MutablePrimArray (PrimState m) Int -> SmallMutableArray (PrimState m) a -> Int -> Int -> a -> Int -> [a] -> m (MutableLoserTree (PrimState m) a, a)
     sift !ids !arr !idxX !j !x !idx0 xs = do
         !idxY <- readPrimArray ids j
         y     <- readSmallArray arr j
+        -- The algorithm has the invariant which states that,
+        -- if idxY < 0 then xs will be non-empty.
```

Comment:
That's not really the invariant, but a conclusion from it. If you put the invariant "length of `xs` is equal to number of unitialised entries (value -1 in `ids`)" as a comment on `sift` itself, you can deduce next to the error call that "entry at j is uninitialised" -> "xs can't be empty".